### PR TITLE
Add media delete and get information API.

### PR DESCRIPTION
**Pull request by dewniMW**
This PR provides the following implementations.
- Add media service delete API.
- Add media service list media information API.
- Move media.properties file related configuration validations to the service layer.
- Remove the existing implementation on reading the media storage type from identity.xml file.
- Read the media storage type from the properties file.

Issue: https://github.com/wso2/product-is/issues/8499

### CODE REVIEW

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java [hunk 1]
```diff
@@ -71,15 +71,17 @@ public String addFile(List<InputStream> inputStream, MediaMetadata mediaMetadata
      * @param tenantDomain The tenant domain of the service call.
      * @param type         The high level content-type of the resource (if media content-type is image/png then
      *                     type would be image).
+     * @param storageType  The media storage type.
      * @return requested file.
      * @throws StorageSystemException Exception related to retrieving the media.
      */
-    public DataContent readContent(String id, String tenantDomain, String type) throws StorageSystemException {
+    public DataContent readContent(String id, String tenantDomain, String type, String storageType) throws
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java:78]: cant we inline 78 and 79

File: components/org.wso2.carbon.identity.media.core/src/test/java/org/wso2/carbon/identity/media/core/StorageSystemManagerTest.java [hunk 2]
```diff
@@ -72,58 +73,6 @@ public void setUp() {
                         databaseBasedStorageSystemFactory);
     }
 
-    @Test
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/test/java/org/wso2/carbon/identity/media/core/StorageSystemManagerTest.java:75]: why were these removed?

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/Util.java [hunk 3]
```diff
@@ -232,4 +238,37 @@ private static String getContext(String endpoint) {
...
+            }
+        } catch (IOException e) {
+            MediaServiceConstants.ErrorMessage errorMessage =
+                    MediaServiceConstants.ErrorMessage.ERROR_CODE_ERROR_LOADING_ALLOWED_CONTENT_TYPES;
+            Response.Status status = Response.Status.INTERNAL_SERVER_ERROR;
+            throw handleException(e, errorMessage, LOG, status, MEDIA_PROPERTIES_FILE_NAME);
+        }
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("MediaStoreType is not configured in media.properties file. Hence proceeding with the " +
+                    "default value.");
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/Util.java:269]: shall we log the default value also?

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java [hunk 4]
```diff
@@ -314,4 +323,134 @@ private void validateAllowedMediaTypes(String mediaTypePathParam) {
...
+     * @param id   Unique identifier for the requested media file to be deleted.
+     */
+    public void deleteMedia(String type, String id) {
+
+        validateAllowedMediaTypes(type);
+        StorageSystemManager storageSystemManager = getStorageSystemManager();
+        String tenantDomain = getTenantDomainFromContext();
+        String mediaStoreType = getMediaStoreType();
+        try {
+            boolean mediaDeleted = storageSystemManager.isMediaDeleted(id, type, tenantDomain, mediaStoreType);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java:340]: does isMediaDeleted responsible for deleting the media as well? if so, let's change the method name to be more meaningful

File: components/org.wso2.carbon.identity.media.core/src/test/java/org/wso2/carbon/identity/media/core/StorageSystemManagerTest.java (hunk 2 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.core/src/test/java/org/wso2/carbon/identity/media/core/StorageSystemManagerTest.java:75]: these tests were removed as the media storage type will no longer be read from the identity.xml

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java (hunk 4 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java]: yes it is responsible for deleting media. Is it ok to change the name to `deleteMedia` ?

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java [hunk 5]
```diff
@@ -240,19 +242,39 @@ public InputStream transform(String id, String type, String tenantDomain, InputS
...
+     * @param mediaTypePathParam The media type available as a path parameter in the upload request.
+     * @param contentSubType     The  subtype of the uploaded media (if media content-type is image/png then
+     *                           subtype would be png).
+     */
+    public void validateFileUploadMediaTypes(String mediaTypePathParam, String contentSubType)
+            throws StorageSystemException {
+
+        HashMap<String, List<String>> allowedContentTypes = StorageSystemUtil.getContentTypes();
+        if (MapUtils.isEmpty(allowedContentTypes) || !allowedContentTypes.keySet().contains(mediaTypePathParam)) {
+            throw new StorageSystemException();
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java:257]: can we add a descriptive error message to this exception?

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java [hunk 6]
```diff
@@ -240,19 +242,39 @@ public InputStream transform(String id, String type, String tenantDomain, InputS
...
+        }
 
-        String contentStoreType = IdentityUtil.getProperty("ContentStore.Type");
-        if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug(String.format("The configured content store type is %s.", contentStoreType));
+        List<String> allowedContentSubTypes = allowedContentTypes.get(mediaTypePathParam);
+        if (CollectionUtils.isEmpty(allowedContentSubTypes) ||
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/StorageSystemManager.java:261]: inline 261 and 262

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/util/StorageSystemUtil.java [hunk 7]
```diff
@@ -15,15 +15,82 @@
...
+                    mediaStorageType = mediaStoreTypeConfig;
+                } else {
+                    if (LOGGER.isDebugEnabled()) {
+                        LOGGER.debug("MediaStoreType is not configured in media.properties file. Hence proceeding" +
+                                " with the default value: " + DEFAULT_MEDIA_STORE_TYPE_VALUE);
+                    }
+                    mediaStorageType = DEFAULT_MEDIA_STORE_TYPE_VALUE;
+                }
+
+                String[] allowedContentTypes = properties.getProperty(ALLOWED_CONTENT_TYPES).split(",");
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/util/StorageSystemUtil.java:86]: can properties.getProperty throw NPE?

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/util/StorageSystemUtil.java [hunk 8]
```diff
@@ -15,15 +15,82 @@
...
+                for (String contentType : allowedContentTypes) {
+                    contentTypes.put(contentType, Arrays.asList(properties.getProperty(contentType +
+                            ALLOWED_CONTENT_SUB_TYPES).split(",")));
+                }
+            }
+        } catch (IOException e) {
+            LOGGER.error("Error while loading media.properties file.", e);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/util/StorageSystemUtil.java:93]: don't we need to throw this exception and break the flow?

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/util/StorageSystemUtil.java (hunk 7 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/util/StorageSystemUtil.java:86]: no as a properties object has been created at https://github.com/wso2/identity-media/pull/20/files/ab606ecd52bd374cf0eee67749e63626e1b33dca#diff-bbbfd5f8451f82948410932d42d5a79cR66

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/util/StorageSystemUtil.java (hunk 8 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/util/StorageSystemUtil.java]: fixed with 8e91b75f7582d8b0694962e84f68ab40e81b2cf8

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/internal/MediaServiceComponent.java [hunk 9]
```diff
@@ -45,13 +46,23 @@
...
-        bundleContext.registerService(StorageSystemFactory.class.getName(), fileBasedStorageSystemFactory, null);
-        StorageSystemFactory dbBasedStorageSystemFactory = new DatabaseBasedStorageSystemFactory();
-        bundleContext.registerService(StorageSystemFactory.class.getName(), dbBasedStorageSystemFactory, null);
-        bundleContext.registerService(StorageSystemManager.class, new StorageSystemManager(), null);
+        try {
+            // Load configurations from media.properties file into memory.
+            StorageSystemUtil.loadMediaProperties();
+
+            BundleContext bundleContext = componentContext.getBundleContext();
+            MediaServiceDataHolder.getInstance().setBundleContext(bundleContext);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/internal/MediaServiceComponent.java:54]: why do we nede to store the bundle context?

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/model/MediaInformation.java [hunk 10]
```diff
@@ -26,7 +26,7 @@
 public class MediaInformation {
 
     private List<String> links;
-    private MediaInformationMetadata mediaInformationResponseMetadata;
+    private Object mediaMetadata;
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/model/MediaInformation.java:29]: why we use Object here? can't we define a specific type/

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/Util.java [hunk 11]
```diff
@@ -56,8 +56,16 @@ private Util() {
      */
     public static StorageSystemManager getStorageSystemManager() {
 
-        return (StorageSystemManager) PrivilegedCarbonContext.getThreadLocalCarbonContext()
-                .getOSGiService(StorageSystemManager.class, null);
+        try {
+            return (StorageSystemManager) PrivilegedCarbonContext.getThreadLocalCarbonContext()
+                    .getOSGiService(StorageSystemManager.class, null);
+        } catch (Throwable e) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/Util.java:62]: what is the requirement to catch throwable here?

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/model/MediaInformation.java (hunk 10 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/model/MediaInformation.java:29]: An object was used as we have decided not to process the json object read from the metadata file. All the metadata stored in the file should be returned without any processing.

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/Util.java (hunk 11 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/Util.java:62]: throwable was used here to handle NPE caused in the getOSGiService method. This is cause we are loading the properties file in https://github.com/wso2/identity-media/pull/20/files/db8534c3e3a745f3fd56809c2dc973db897e78be#diff-198e0ade17d75d1c1d73b3b7952e701aR51
If there is an error while reading the file the line: https://github.com/wso2/identity-media/pull/20/files/db8534c3e3a745f3fd56809c2dc973db897e78be#diff-198e0ade17d75d1c1d73b3b7952e701aR59 will not be executed, hence the above NPE which I have mentioned.

File: components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/internal/MediaServiceComponent.java (hunk 9 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.core/src/main/java/org/wso2/carbon/identity/media/core/internal/MediaServiceComponent.java]: removed it

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java [hunk 12]
```diff
@@ -307,11 +250,69 @@ private DataContent downloadMedia(String type, String id) {
...
+     *             type would be image.
+     * @param id   Unique identifier for the requested media file to be deleted.
+     */
+    public void deleteMedia(String type, String id) {
+
+        validateAllowedMediaTypes(type);
+        StorageSystemManager storageSystemManager = getStorageSystemManager();
+        String tenantDomain = getTenantDomainFromContext();
+        try {
+            boolean mediaDeleted = storageSystemManager.isMediaDeleted(id, type, tenantDomain);
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java:275]: Is this a media existence check? If so we normally follow `isMediaExist` format.

Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java:275]: this method would delete the media and return true if the delete has been successful

Comment (ashensw) [reply components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java:275]: The name will give an incorrect meaning IMO. Shall we change the delete method as void and throw an exception if something goes wrong during the deletion process. Then catch that exception and handle it from here accordingly.

Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java]: fixed with: 0d72742b085d7a3aa6c70ce0860be0a1580b9715

**Review by ashensw — APPROVED** [review]

**Review by emswbandara — APPROVED** [review]
